packages/core/src/sdk/EventMappers/EventMapper.ts

```diff
@@ -0,0 +1,94 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+import { InternalLog } from '../../InternalLog';
+import { SdkVerbosity } from '../../SdkVerbosity';
+import { DdSdk } from '../../foundation';
+import { AttributesSingleton } from '../AttributesSingleton/AttributesSingleton';
+import type { Attributes } from '../AttributesSingleton/types';
+import { UserInfoSingleton } from '../UserInfoSingleton/UserInfoSingleton';
+import type { UserInfo } from '../UserInfoSingleton/types';
+
+import { deepClone } from './utils/deepClone';
+
+export type AdditionalEventDataForMapper = {
+    userInfo: UserInfo;
+    attributes: Attributes;
+};
+
+/**
+ * Generic class for applying event mappers.
+ *
+ * Calls params in this order: formatRawEventForMapper, eventMapper, formatMapperEventForNative.
+ *
+ * @param eventMapper the user-registered event mapper
+ * @param formatRawEventForMapper formatter that gets the raw event (from javascript call) and returns the input given to the mapper
+ * @param formatMapperEventForNative formatter that gets the ouput of the mapper, and returns the input given to the native SDKs
+ * @param formatRawEventForNative called when no event mapper is registered
+ */
+export class EventMapper<RawEvent, MapperEvent, NativeEvent> {
+    private eventMapper?: (event: MapperEvent) => MapperEvent | null;
+    private formatRawEventForMapper: (
+        event: RawEvent,
+        additionalData: AdditionalEventDataForMapper
+    ) => MapperEvent;
+    private formatMapperEventForNative: (
+        event: MapperEvent,
+        originalEvent: MapperEvent
+    ) => NativeEvent;
+    private formatRawEventForNative: (event: RawEvent) => NativeEvent;
+
+    constructor(
+        eventMapper: ((event: MapperEvent) => MapperEvent | null) | undefined,
+        formatRawEventForMapper: (
+            event: RawEvent,
+            additionalData: AdditionalEventDataForMapper
+        ) => MapperEvent,
+        formatMapperEventForNative: (
+            event: MapperEvent,
+            originalEvent: MapperEvent
+        ) => NativeEvent,
+        formatRawEventForNative: (event: RawEvent) => NativeEvent
+    ) {
+        this.eventMapper = eventMapper;
+        this.formatRawEventForMapper = formatRawEventForMapper;
+        this.formatMapperEventForNative = formatMapperEventForNative;
+        this.formatRawEventForNative = formatRawEventForNative;
+    }
+
+    public applyEventMapper = (rawEvent: RawEvent): NativeEvent | null => {
+        if (!this.eventMapper) {
+            return this.formatRawEventForNative(rawEvent);
+        }
+
+        // formatting
+        const userInfo = UserInfoSingleton.getInstance().getUserInfo();
+        const attributes = AttributesSingleton.getInstance().getAttributes();
```

Comment:
we are going to move away from singletons in SDK v2, so not sure we should rely on them with the new code developed.

Reply:
These singleton are in JS only, so this will work with multiple instances of the Native SDK initialized.
I'm not sure we'll have to support the scenario of multiple instances of RN SDK initialized, and in case we don't have to I think this saves us a bit of work.
If we ever want to support multiple RN instances, I think it will require a lot more work anyway so it will be a dedicated epic, so I believe it's ok to skip it for now.

Reply:
This is a valid point, but I'm afraid we have to rely on a Singleton pattern in one way or another here, as `setUser` or `setAttributes` can be called before or after `initialize`, we need to make sure we initialize the provider only once and always reference the correct one.